packages/core-data/src/entities.ts

```diff
@@ -136,33 +188,125 @@ export const defaultEntities = [
 		label: __( 'Navigation Area' ),
 		key: 'name',
 		getTitle: ( record ) => record?.description,
-	},
-	{
+	} as const ),
+	defineEntity( {
 		label: __( 'Global Styles' ),
 		name: 'globalStyles',
 		kind: 'root',
 		baseURL: '/wp/v2/global-styles',
 		baseURLParams: { context: 'edit' },
 		plural: 'globalStylesVariations', // should be different than name
 		getTitle: ( record ) => record?.title?.rendered || record?.title,
-	},
-	{
+	} as const ),
+	defineEntity( {
 		label: __( 'Themes' ),
 		name: 'theme',
 		kind: 'root',
 		baseURL: '/wp/v2/themes',
 		baseURLParams: { context: 'edit' },
 		key: 'stylesheet',
-	},
-	{
+	} as const ),
+	defineEntity( {
 		label: __( 'Plugins' ),
 		name: 'plugin',
 		kind: 'root',
 		baseURL: '/wp/v2/plugins',
 		baseURLParams: { context: 'edit' },
 		key: 'plugin',
-	},
-];
+	} as const ),
+] as const;
+
+type Element = typeof defaultEntities[ number ];
+
+type DefaultEntityContext<
+	K extends string,
+	N extends string,
+	E = Element & {
+		kind: K;
+		name: N;
+	}
+> = 'baseURLParams' extends keyof E
+	? 'context' extends keyof E[ 'baseURLParams' ]
+		? E[ 'baseURLParams' ][ 'context' ]
+		: 'view'
+	: 'view';
+
+type EntityKeyName<
+	K extends string,
+	N extends string,
+	E = Element & {
+		kind: K;
+		name: N;
+	}
+> = 'key' extends keyof E ? E[ 'key' ] : 'id';
+
+type DeclaredEntity< K extends string, N extends string, RecordType > = {
+	kind: K;
+	name: N;
+	recordType: RecordType;
+	keyType: EntityKeyName< K, N > extends keyof RecordType
+		? RecordType[ EntityKeyName< K, N > ]
+		: never;
+	defaultContext: DefaultEntityContext< K, N >;
+};
+type APIEntity<
+	K extends string,
+	N extends string,
+	RecordType,
+	KeyName = 'id',
+	C = 'edit'
+> = {
+	kind: K;
+	name: N;
+	recordType: RecordType;
+	keyType: KeyName extends keyof RecordType ? RecordType[ KeyName ] : never;
+	defaultContext: C;
+};
+
+export type EntityTypeUnion< C extends Context = any > =
```

Comment:
nitpick: we don't need `Union` in the name. Could be something like `EntityTypeDefinition` or `DeclaredEntity` or `EntityDeclaration`